CMakeLists.txt

```diff
@@ -151,7 +151,40 @@ if(BML_MPI)
   add_definitions(-DDO_MPI)
 endif()
 
-set(BML_CUDA FALSE CACHE BOOL "Whether to compile with CUDA support")
+option(BML_GPU "Compile with OpenMP GPU Offload support" FALSE)
+if(BML_GPU)
+  message(STATUS "Will build with OpenMP GPU Offload support")
+  add_definitions(-DDO_GPU)
+
+  find_package(OpenMP 4.5 REQUIRED)
+  set(OpenMP_CXX_VERSION ${OpenMP_CXX_VERSION} CACHE STRING "OpenMP CXX Version")
+
+  include(CheckCXXCompilerFlag)
+  if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+    set(OpenMP_CXX_OFFLOAD_FLAG "-foffload='-lm'")
+  elseif(MAKE_CXX_COMPILER_ID STREQUAL "Intel")
+    set(OpenMP_CXX_OFFLOAD_FLAG "-foffload='-lm'")
+  elseif(MAKE_CXX_COMPILER_ID STREQUAL "Clang")
+    set(OpenMP_CXX_OFFLOAD_FLAG "-fopenmp-targets=nvptx64-nvidia-cuda --cuda-path=${CUDA_TOOLKIT_ROOT_DIR}")
+  endif()
+  check_cxx_compiler_flag("${OpenMP_CXX_OFFLOAD_FLAG}" OFFLOAD_CXX_FOUND)
+  if(OFFLOAD_CXX_FOUND)
+    message("Found offload")
+  endif()
```

Comment:
Should there be an error if offload doesn’t work?

Reply:
That I don't know. How does this work in OpenMP? Will it silently fall back to running on the host if it can't offload to a device?

Reply:
No, I meant, if the compiler doesn't support `${OpenMP_CXX_OFFLOAD_FLAG}` it shouldn't be added to the flags below.

Reply:
Oh, good point. I agree with that :smile: